### PR TITLE
Add a generic workflow nyuki using 'tukio' and 'motor' librairies on top of mongo

### PR DESCRIPTION
**Pull request by gdraynz**

### CODE REVIEW

File: nyuki/meta/workflow/workflow.py [hunk 1]
```diff
@@ -0,0 +1,495 @@
...
+            return Response({
+                **tmpl_dict,
+                **metadata,
+                'errors': self.errors_from_validation(template)
+            })
+
+    @resource(r'/templates/{tid}', version='v1')
+    class Template:
+
+        async def get(self, request, tid):
```
Comment (lerignoux) [inline nyuki/meta/workflow/workflow.py]: this shouldn't returns a list !

File: nyuki/meta/workflow/workflow.py [hunk 2]
```diff
@@ -0,0 +1,495 @@
...
+            Return workflow instances
+            """
+            return Response(
+                json.dumps(
+                    [wflow for wflow in self.engine.instances.values()],
+                    default=self.serialize_wflow_exec),
+                content_type='application/json'
+            )
+
+        async def put(self, request):
```
Comment (lerignoux) [inline nyuki/meta/workflow/workflow.py]: I got an error 500 when calling this route:
```
21 15:37:08,652  INFO     [nyuki.meta.workflow.storage] Update draft for query: {'id': 'b5af58a3-ef0c-4b63-b55a-3443e86a4af0', 'draft': True}
2016-06-21 15:41:49,854  ERROR    [nyuki.bus.reporting] Traceback (most recent call last):
  File "/home/laurent/Optiflows/src/nyuki/nyuki/api.py", line 150, in middleware
    capa_resp = await capa_handler(request, **request.match_info)
  File "/home/laurent/.pyenv/versions/3.5.0/lib/python3.5/asyncio/coroutines.py", line 212, in coro
    res = yield from await_meth()
  File "/home/laurent/Optiflows/src/nyuki/nyuki/meta/workflow/workflow.py", line 441, in put
    wflow = self.engine.run(request['id'], request.get('inputs', {}))
  File "/home/laurent/.pyenv/versions/test/lib/python3.5/site-packages/tukio/engine.py", line 314, in run
    wflow = new_workflow(template, loop=self._loop)
  File "/home/laurent/.pyenv/versions/test/lib/python3.5/site-packages/tukio/workflow.py", line 660, in new_workflow
    policy_handler = OverrunPolicyHandler(wf_tmpl, loop=loop)
  File "/home/laurent/.pyenv/versions/test/lib/python3.5/site-packages/tukio/workflow.py", line 94, in __init__
    self.policy = template.policy
AttributeError: 'str' object has no attribute 'policy'

2016-06-21 15:41:49,854  ERROR    [nyuki.api] Error handling request
Traceback (most recent call last):
  File "/home/laurent/Optiflows/src/nyuki/nyuki/api.py", line 150, in middleware
    capa_resp = await capa_handler(request, **request.match_info)
  File "/home/laurent/.pyenv/versions/3.5.0/lib/python3.5/asyncio/coroutines.py", line 212, in coro
    res = yield from await_meth()
  File "/home/laurent/Optiflows/src/nyuki/nyuki/meta/workflow/workflow.py", line 441, in put
    wflow = self.engine.run(request['id'], request.get('inputs', {}))
  File "/home/laurent/.pyenv/versions/test/lib/python3.5/site-packages/tukio/engine.py", line 314, in run
    wflow = new_workflow(template, loop=self._loop)
  File "/home/laurent/.pyenv/versions/test/lib/python3.5/site-packages/tukio/workflow.py", line 660, in new_workflow
    policy_handler = OverrunPolicyHandler(wf_tmpl, loop=loop)
  File "/home/laurent/.pyenv/versions/test/lib/python3.5/site-packages/tukio/workflow.py", line 94, in __init__
    self.policy = template.policy
AttributeError: 'str' object has no attribute 'policy'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/laurent/.pyenv/versions/test/lib/python3.5/site-packages/aiohttp/server.py", line 266, in start
    yield from self.handle_request(message, payload)
  File "/home/laurent/.pyenv/versions/test/lib/python3.5/site-packages/aiohttp/web.py", line 90, in handle_request
    resp = yield from handler(request)
  File "/home/laurent/Optiflows/src/nyuki/nyuki/api.py", line 155, in middleware
    reporting.exception(exc)
  File "/home/laurent/Optiflows/src/nyuki/nyuki/bus/reporting.py", line 167, in exception
    self._loop.call_later(
AttributeError: 'NoneType' object has no attribute 'call_later'
```

File: nyuki/meta/workflow/workflow.py [hunk 3]
```diff
@@ -0,0 +1,495 @@
...
+
+            return Response(
+                json.dumps(wflow, default=self.serialize_wflow_exec),
+                content_type='application/json'
+            )
+
+        async def post(self, request, iid):
+            """
+            Operate on a workflow instance
+            """
```
Comment (lerignoux) [inline nyuki/meta/workflow/workflow.py]: what about raise NotImplementedError ?

File: nyuki/meta/workflow/workflow.py [hunk 4]
```diff
@@ -0,0 +1,495 @@
...
+            return Response(
+                json.dumps(
+                    [wflow for wflow in self.engine.instances.values()],
+                    default=self.serialize_wflow_exec),
+                content_type='application/json'
+            )
+
+        async def put(self, request):
+            """
+            Start a workflow from a template in payload
```
Comment (lerignoux) [inline nyuki/meta/workflow/workflow.py]: that start a draft workflow isn't it ?
not sure of the behaviour could we talk about it ?

File: nyuki/meta/workflow/workflow.py [hunk 5]
```diff
@@ -0,0 +1,495 @@
...
+
+        async def get(self, request):
+            """
+            Return the list of available topics
+            """
+            return Response(self.topics)
+
+    @resource('/test', version='v1')
+    class Test:
+
```
Comment (lerignoux) [inline nyuki/meta/workflow/workflow.py]: same exception than on instance

File: nyuki/meta/workflow/workflow.py [hunk 6]
```diff
@@ -0,0 +1,495 @@
...
+    class Tasks:
+
+        async def get(self, request):
+            """
+            Return the available tasks
+            """
+            return Response(self.AVAILABLE_TASKS)
+
+    @resource('/topics', version='v1')
+    class Topics:
```
Comment (lerignoux) [inline nyuki/meta/workflow/workflow.py]: always returns (in spite of having a production template with topics)

File: nyuki/meta/workflow/workflow.py (hunk 2 above)
Comment (gdraynz) [inline nyuki/meta/workflow/workflow.py]: Please check your dependencies, I didn't modify any of the current nyuki-workflow code

File: nyuki/meta/workflow/workflow.py (hunk 6 above)
Comment (gdraynz) [inline nyuki/meta/workflow/workflow.py]: This route returns the list of topics set in the configuration file. If no topic if given, the list is empty

Comment (lerignoux) [inline nyuki/meta/workflow/workflow.py]: ok

File: nyuki/meta/workflow/workflow.py (hunk 2 above)
Comment (lerignoux) [inline nyuki/meta/workflow/workflow.py]: fixed in NYWF-25

File: nyuki/meta/workflow/workflow.py (hunk 5 above)
Comment (lerignoux) [inline nyuki/meta/workflow/workflow.py]: same fix

**Pull request merged**